all-sdk/autotuning.md

```diff
@@ -0,0 +1,431 @@
+In Core: `max_concurrent_wft_polls` and `max_concurrent_at_polls`. There's no need for this number to be fixed, and we
+can and should scale it up and down as demand dictates.  This work is largely already designed and implemented in Go:
+[see the branch
+here](https://github.com/temporalio/sdk-go/compare/master...Quinn-With-Two-Ns:sdk-go:poller-autottune-v0.2). Hence this
+doc will not focus on this area in detail.
+
+It is important to note that there is a distinction between the active number of poller *routines* and the active number
+of *open actual RPC calls*. The number of routines may be scaled independently of the number of open RPC calls at any
+time. The former may have a floor, whereas the latter must always be able to go to 0. Users are unlikely to care about
+the number of routines as they are relatively cheap. They *do*, however, care about how quickly we will change the
+number of active RPCs. The floor on the number of routines effectively determines how quickly we are willing to consume
+available slots and thus make actual RPC calls. IE: If the floor on the number of routines is 10, and 10 slots open up
+at the same time, we will immediately make 10 RPC calls. 
+
+In Go the name for these options is already misleading: `MaxConcurrentWorkflowTaskPollers` is not accurate, it's really
+`NumWorkflowTaskPollers` (or routines), what it probably should be is `MaxConcurrentWorkflowTaskPolls`, since the number
+of open RPC calls is actually what's fluctuating.
+
+
+### Slot tuning & Cache Size
+
+Workers have a configurable number of slots for task types. In Core: `max_outstanding_workflow_tasks`,
+`max_outstanding_activities`, and `max_outstanding_local_activities`. These are set by the user and can't change at
+runtime. Autotuning of these values seeks to allow them to change at runtime according to available resources on
+the worker. It is not feasible for us to choose *optimal* values for these at runtime, because we cannot know a-priori
+about how much resources any given workflow or activity task will consume - only the user can (maybe) know this.
+However, we can provide reasonable default autotuning implementations while also allowing users to provide their own.
+
+Additionally, there is a maximum cache size for cached workflows. In Core: `max_cached_workflows`. This is also set by
+users at the moment, and autotuning it is subject to the same lack of knowledge we have as with slot tuning.
+
+## Constraints
+
+There are some invariants that must be maintained:
+
+* The current number of active polls for a task type must be <= the number of free slots for that task type at all
+  times. Otherwise, we can end up receiving a task that won't have anywhere to go until a slot frees. For example,
+  in Core this is expressed by using a permit reservation system.
+* Max outstanding workflow tasks must be <= max cached workflows.
```

Comment:
I don't think this is true? You can run a worker with no cache today

Reply:
Yeah that's true - but practically speaking it means your cache's peak size will equal the number of concurrent workflow tasks. In a real-world setting you probably care about cache size as it relates to peak memory usage, and since users are (should) always using the cache, the constraint makes sense in that context.